### ISSUE TITLE
Port mapping should be IP:4000:80

### ISSUE THREAD

**Issue by behemphi** [jekyll/docker#63]
While looking at #60, I did not get a response from the server. As best I can tell, within the container, `nginx` is listening on port 80.  So, I believe the run command from the README should read:

```
docker run --rm --label=jekyll --volume=$(pwd):/srv/jekyll \
  -it -p 127.0.0.1:4000:80 jekyll/jekyll jekyll s
```

or 

```
docker run --rm --label=jekyll --volume=$(pwd):/srv/jekyll \
  -it -p $(docker-machine ip `docker-machine active`):4000:80 \
    jekyll/jekyll jekyll s
```

**Comment by envygeeks** [comment]
You'll need to provide way more context because that's a negative.  We listen on both ports 4000 and 80.  WEBRick starts on 4000 and we make it available on 0.0.0.0 and Nginx is there for when you need to go advanced stuff.  We will not recommend people block either of them, so if you are having a problem provide more context because that will not change as the default recommendation but we make notes for non-native Docker.

**Comment by envygeeks** [comment]
What I am saying is, Nginx is *not* our current recommend path to getting to Jekyll in our Docker images, you hitting port 4000 *is* our recommended path, port 80 (nginx) only exists because somebody wished for us to behave like Github pages on the pages tag, so I did as they wished as I said I would if I could for the pages tag, then I added stuff so people could do advanced stuff.

It could very well end up in the near future that Nginx is removed entirely because I recently committed to master the same behaviors that Github pages have (/page routing to /page.html if /page/index.html does not exist) so we no longer need Ngninx, and I also have a pull for core review that enables a default charset further addressing the need for Nginx in development.

Ultimately though, the recommendation right now is for you to use port 4000:4000 (which is why it sits on 0.0.0.0 instead of localhost) and Nginx only if you need it for reasons we are not willing to cover as corners and edges.

**Comment by behemphi** [comment]
Thanks again for the quick and detailed response @envygeeks!

I am not suggesting the port set up change.  I am saying that the command, as provided in the README and by @parkr in #60 did not work for me.  I am on a Mac using toolbox with 1.9.1.  

In the short of it, if I use this command: 

```
docker run --rm --label=jekyll --volume=$(pwd):/srv/jekyll \
  -it -p 127.0.0.1:4000:80 jekyll/jekyll jekyll s
```

I get no response from the browser. However if I use the suggested command (change to port 80 internal to the container), I get my site served in the browser. 

---

### TL;DR

After the happened to me a couple of times I got in the the container thus:

`docker exec -it parkr sh`  (the containers name being parkr)

I checked the process list and saw this:

```
/srv/jekyll # behemphi:~/behemphi/behemphi.github.io 21:54:30 > docker exec -it parkr sh
/srv/jekyll # ps -ef
PID   USER     TIME   COMMAND
    1 root       0:00 {startup} /bin/sh /usr/local/bin/startup jekyll build --watch
   13 root       0:00 runsvdir -P /etc/startup2.d
   14 root       0:00 runsvdir -P /etc/startup3.d
   16 jekyll     0:00 ruby /usr/lib/ruby/gems/2.2.0/gems/jekyll-2.5.3/bin/jekyll build --watch --future --unpublished
   18 root       0:00 runsv nginx
   19 root       0:00 nginx: master process nginx
   20 nginx      0:00 nginx: worker process
   21 nginx      0:00 nginx: worker process
   22 nginx      0:00 nginx: worker process
   23 nginx      0:00 nginx: worker process
   36 jekyll     0:00 /bin/sh -c /usr/bin/python2 /usr/lib/ruby/gems/2.2.0/gems/pygments.rb-0.6.3/lib/pygments/mentos.py
   38 jekyll     0:01 /usr/bin/python2 /usr/lib/ruby/gems/2.2.0/gems/pygments.rb-0.6.3/lib/pygments/mentos.py
   42 root       0:00 sh
   47 root       0:00 ps -ef
```

So, I don't see anything like a `jekyll serve` command (remember I am totally new to Jekyll) and I did see `nginx`.  

Checking `/etc/nginx/conf.d/jekyll.conf I see that `nginx` is, indeed, listening on 80.  

Checking ports, it appears only nginx is listening (inside the container):

```
/srv/jekyll # netstat -p
Active Internet connections (w/o servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
Active UNIX domain sockets (w/o servers)
Proto RefCnt Flags       Type       State         I-Node PID/Program name    Path
unix  3      [ ]         STREAM     CONNECTED      80529 20/nginx: master pr 
unix  3      [ ]         STREAM     CONNECTED      80526 20/nginx: master pr 
unix  3      [ ]         STREAM     CONNECTED      80523 20/nginx: master pr 
unix  3      [ ]         STREAM     CONNECTED      80530 20/nginx: master pr 
unix  3      [ ]         STREAM     CONNECTED      80524 20/nginx: master pr 
unix  3      [ ]         STREAM     CONNECTED      80528 20/nginx: master pr 
unix  3      [ ]         STREAM     CONNECTED      80527 20/nginx: master pr 
unix  3      [ ]         STREAM     CONNECTED      80525 20/nginx: master pr 
```

In issue #60, this is what lead me to think that I needed to run both `jekyll build` and `jekyll server` as different processes to achieve an interactive development environment.

**Comment by behemphi** [comment]
After more carefully reading your comment @envygeeks, I realized that I dropped the image tag. I edited the comment above, but for clarity, I am using jekyll:jekyll:pages.

**Comment by envygeeks** [comment]
That's the problem, I thought we fixed that damn bug.  There was a bug in Jekyll where jekyll s didn't work the way it was supposed to work, it didn't boot the server at all. We attempted to work around it by removing the "s" and making sure it was serve, but it seems apparent I either have a race by order in my shim or something is reshimming my shim.

**Comment by envygeeks** [comment]
This is what you should see when you do `jekyll s` (this is me using the master image)

```
PID   USER     TIME   COMMAND
    1 root       0:00 {startup} /bin/sh /usr/local/bin/startup jekyll s --trace
   36 root       0:00 runsvdir -P /etc/startup2.d
   37 root       0:00 runsvdir -P /etc/startup3.d
   39 root       0:00 runsv nginx
   40 jekyll     0:01 ruby /usr/lib/ruby/gems/2.2.0/gems/jekyll-3.0.1/bin/jekyll serve -H 0.0.0.0 --trace --future --unpublished
   41 root       0:00 nginx: master process nginx
   43 nginx      0:00 nginx: worker process
   44 nginx      0:00 nginx: worker process
   45 nginx      0:00 nginx: worker process
   46 nginx      0:00 nginx: worker process
   67 root       0:00 sh
   72 root       0:00 ps auxf
```

However, yours is telling it to do Jekyll build, instead of Jekyll serve, when you do `s` are you sure you are running `jekyll s` and not `jekyll b` or `jekyll build`?

**Comment by envygeeks** [comment]
Actually, I'm inclined to believe you did accidently do `jekyll build` instead of `jekyll s` or `jekyll serve` because PID one shows exactly that command because we display the command we are given.

**Comment by envygeeks** [comment]
So, I reread that ticket and read this ticket.  I see the problem, there you were using Jekyll Build, you still are using Jekyll Build, we only enable port 4000 on Jekyll Serve, if you are in active development you should be using `jekyll s` if you wish to have a server as `jekyll b` does not boot up a server, it only boot up Jekyll to watch (if you ask it to) and build.  So if you explictly decide not to boot up the server yes you will need to hit port 80, but if you boot up `jekyll s` properly for development, it will kick on 4000.

**Comment by envygeeks** [comment]
I'll leave this open until you confirm or deny that note solves your problem.

**Comment by behemphi** [comment]
Thanks @envygeeks 

My goal is to have the container run a build watch and server (at the same time) so that I can save a blog post on my local workstation and see the change served via the container run time. Thus I think I am at a bit of an impasse.  

So before I waste anymore of your time, is my goal clear and sensible?

Again, really appreciate your quick/detailed responses.

**Comment by envygeeks** [comment]
If you want to build and serve then use `jekyll serve` with `4000:4000` (my recommended path,) if you wish to simply just build, and then serve it yourself then use `jekyll build --watch` and you can choose your own path, you can serve it via your host Nginx or you can use Docker's Nginx.

Both `jekyll serve` and `jekyll build` will respond to the environment variables created for #60.
